docs: signals and actions documentation (#6641)
Unleash/unleash: website/docs/reference/actions.mdx

---
title: Actions
---

:::info Availability

> Unleash actions is an **enterprise-only** feature released as beta in Unleash v5.11.

Reviewer: Actions were introduced as a beta feature in **Unleash 5.11** and are only available in Unleash Enterprise. We plan to make this feature generally available to all Enterprise users in **Unleash 5.12**.

> Unleash actions is an **enterprise-only** feature released as beta in Unleash v5.11.

Reviewer: ```suggestion
Actions were introduced as a beta feature in **Unleash 5.11** and are only available in Unleash Enterprise. We plan to make this feature generally available to all Enterprise users in **Unleash 5.12**.
```

:::
## Overview

> Actions enable our users to execute automated actions based on signals received from an external system.

Reviewer: ```suggestion
Actions allow you to configure automated actions to be executed based on signals received from an external system.
```

![Actions executed by Unleash.](/img/signals/action-execution.png)

## Creating actions

:::info Permissions

> During the beta phase, creating actions require admin permissions.

Reviewer: I also assume this will be a project-level permission in the future, but I think for now it's safer to just mention this.
```suggestion
Creating actions requires the `ADMIN` permission.
```

:::

Actions are scoped to a project. To be able to create actions, you need to navigate to the project settings and find the **Actions** tab.

![Actions tab.](/img/signals/actions-tab.png)

### Name and description
When creating a new action you first need to specify a name and a description. This will help you identify your action and give it a meaningful description.

### Source and filters

> Then you need to select a source. If you haven't created a signal endpoint yet, you can do so by clicking the **Create new signal endpoint** button (more information on this in the [Signals](./signals.mdx) section).

Reviewer: ```suggestion
Then you need to select a source. If you haven't created a signal endpoint yet, you can do so by using the **Create new signal endpoint** link (more information on this in the [Signals](./signals.mdx) section).
```

> Once selected the source, you'll be able to see a preview of the latest signal which will help you to define different filters that need to match before executing the action.

Reviewer: ```suggestion
If the source you selected already received at least one signal, you'll be able to see a preview of the latest signal received. This can help you define the different filters that need to match for the action to be executed.
```

> Filters work similar as [feature flag strategy constraints](./strategy-constraints.md) where the signal payload acts as the context for the constraint evaluation.

Reviewer: ```suggestion
Filters work similarly to [feature flag strategy constraints](./strategy-constraints.md), where the signal payload acts as the context for the constraint evaluation.
```

### Actions

> When Unleash executes actions as reaction to an external signal, it will use a [service account](./service-accounts.md) to execute them. This service account needs to have the necessary permissions to execute the actions. This service account doesn't need an API token, so when creating it, don't generate an API token and give it the necessary permissions only.

Reviewer 2: I'm not sure that link would work?

Reviewer: Seems to work for me: https://unleash-docs-git-signals-docs-unleash-team.vercel.app/reference/actions

> When Unleash executes actions as reaction to an external signal, it will use a [service account](./service-accounts.md) to execute them. This service account needs to have the necessary permissions to execute the actions. This service account doesn't need an API token, so when creating it, don't generate an API token and give it the necessary permissions only.

Reviewer: ```suggestion
When these actions are triggered, they will execute under the identity of a [service account](./service-accounts.md). This means that when you later check the [event log](./event-log) you will see this service account as the actor for those events. In order to execute the configured actions successfully, this service account needs to have the necessary permissions. This service account doesn't need an API token, so when creating it you can skip the token generation step and simply make sure it has the necessary permissions.
```

> The actions are executed in the order they are defined. You can add multiple actions to an action set. Depending on the action you want to execute you will have to provide different parameters.

Reviewer: ```suggestion
The actions are executed in the same order that they are defined. If a previous action fails, the following actions will not be started. You can add multiple actions to an action set and you can add multiple action sets to a project. Depending on the action you want to execute you will have to provide different parameters.

If an action 
```

> 

Reviewer: Maybe these sections could use some screenshots.

Reviewer: Addressed in https://github.com/Unleash/unleash/pull/6641/commits/fc401d2e0ba8cd3e2f9aae103eea68f064ae4d85

## Viewing action executions

> Once the action is functional and signals start to match the filters, you can see the action executions in the **Actions** tab.

Reviewer: ```suggestion
Once the action is configured and you start receiving signals that match the filters you configured, you can see the action executions in the **Actions** tab.
```

![Action executions option.](/img/signals/view-events.png)

> Clicking on the **View events** button will show you the details of the action executions (what was the payload, did they execute successfully or not, what was the error in case of an error, etc).

Reviewer: ```suggestion
Using the **View events** option will show you the details of the action executions (what was the payload, did they execute successfully or not, what was the error in case of an error, etc).

To view the action events as they are executed, select the **View events** option in the respective action row. This will show you more information about each action execution: The signal payload, whether the actions executed successfully, the error details in case of an error, and other details.
```

![Event details.](/img/signals/action-execution-log.png)

## Limits that apply

> There are some constraints that apply to the execution of actions which are meant to protect the system from being overloaded. 

Reviewer: ```suggestion
There are some constraints on the actions feature as a measure to protect the stability of Unleash. These values can be overridden for self-hosted installations.
```

> Unprocessed signals are processed in a first in first out manner every 1 minute. If the number of signals to process exceeds the limit, the newest signals will be processed with some lag. 

Reviewer: ```suggestion
Unprocessed signals are processed in a "first in, first out" fashion, once every minute. If the number of pending signals to process exceeds the limit, it may take longer for the latest signals to be processed.
```

The limits are as follows:

- **Max action sets per project**: 5.
- **Max filters per action set**: 5.
- **Max distinct values in a filter**: 25.
- **Max actions per action set**: 10.

> - **Max signals to retrieve per cycle**: 10.

Reviewer: ```suggestion
- **Max signals to process per cycle**: 10.
```

To view the action events as they are executed, select the **View events** option in the respective action row. This will show you more information about each action execution: The signal payload, whether the actions executed successfully, the error details in case of an error, and other details.

![Event details.](/img/signals/action-execution-log.png)

## Limits that apply
There are some constraints on the actions feature as a measure to protect the stability of Unleash. These values can be overridden for self-hosted installations.

Unprocessed signals are processed in a first in first out manner every 1 minute. If the number of signals to process exceeds the limit, the newest signals will be processed with some lag. 

The limits are as follows:

- **Max action sets per project**: 5.
- **Max filters per action set**: 5.
- **Max distinct values in a filter**: 25.
- **Max actions per action set**: 10.
- **Max signals to retrieve per cycle**: 10.

Ref: [Unleash Config](https://github.com/Unleash/unleash/blob/859fe098fedc261d646833012d9d408039491075/src/lib/create-config.ts#L577-L604)

## Sending signals to Unleash
See [Signals](./signals.mdx)

> import { info } from "console"

Reviewer: ```suggestion

```
